src/Database/LSMTree/Internal/Lookup.hs

```diff
@@ -166,20 +186,231 @@ indexSearches !rs !ks !rkixs = do
+  where
+    precondition = and [
+          V.map Run.runFilter rs == blooms
+        , V.map Run.runIndex rs == indexes
+        , V.length rs == V.length kopsFiles
+        ]
+
+{-# SPECIALIZE intraPageLookups ::
+       V.Vector (Run (Handle HandleIO))
+    -> V.Vector SerialisedKey
+    -> VU.Vector (RunIx, KeyIx)
+    -> V.Vector (IOOp RealWorld HandleIO)
+    -> VU.Vector IOResult
+    -> IO (V.Vector (Maybe (Entry SerialisedValue (BlobRef (Run (Handle HandleIO))))))
+  #-}
+-- | Intra-page lookups.
+--
+-- TODO: optimise by reducing allocations, possibly looking at core, using
+-- unsafe vector operations.
+--
+-- TODO: generalised value resolution. We currently use @combineNormal@, but we
+-- should be able to re-use this code for @combineMonoidal@.
+--
+-- PRECONDITION: @length rkixs == length ioops == length ioress@
+intraPageLookups ::
+     forall m h. (PrimMonad m, MonadThrow m)
+  => V.Vector (Run (Handle h))
+  -> V.Vector SerialisedKey
+  -> VU.Vector (RunIx, KeyIx)
+  -> V.Vector (IOOp (PrimState m) h)
+  -> VU.Vector IOResult
+  -> m (V.Vector (Maybe (Entry SerialisedValue (BlobRef (Run (Handle h))))))
+intraPageLookups !rs !ks !rkixs !ioops !ioress =
+    assert precondition $ do
+      res <- VM.replicate (V.length ks) Nothing
+      loop res 0
+      V.unsafeFreeze res
+  where
+    precondition = and [
+          VU.length rkixs == V.length ioops
+        , V.length ioops == VU.length ioress
+        ]
+
+    !n = V.length ioops
+
+    loop ::
+         VM.MVector (PrimState m) (Maybe (Entry SerialisedValue (BlobRef (Run (Handle h)))))
+      -> Int
+      -> m ()
+    loop !res !ioopix
+      | ioopix == n =  pure ()
+      | otherwise = do
+          let ioop = ioops `V.unsafeIndex` ioopix
+              iores = ioress `VU.unsafeIndex` ioopix
+          checkIOResult ioop iores
+          let (rix, kix) = rkixs `VU.unsafeIndex` ioopix
+              r = rs `V.unsafeIndex` rix
+              k = ks `V.unsafeIndex` kix
+          buf <- unsafeFreezeByteArray (ioopBuffer ioop)
+          let boff = unBufferOffset $ ioopBufferOffset ioop
+          case rawPageLookup (makeRawPage buf boff) k of
+            LookupEntryNotPresent -> pure ()
+            -- Laziness ensures that we only compute the forcing of the value in
+            -- the entry when the result is needed.
+            LookupEntry e         -> do
+                let e' = Just (bimap copySerialisedValue (mkBlobRef r) e)
+                unsafeModifyMMaybeStrict res (`combineNormal` e') kix
+            -- Laziness ensures that we only compute the appending of the prefix
+            -- and suffix when the result is needed. We do not use 'force' here,
+            -- since appending already creates a new primary vector.
+            --
+            -- TODO: verify if appending always computes a new primary vector
+            LookupEntryOverflow e m -> do
+                let v' (SerialisedValue v) = SerialisedValue $ v <>
```

Comment:
Ah, so since `combine` often doesn't need to force its second argument, we don't pay for the copy, but if the entry ends up in `res` (i.e. potentially lives long), it is forced, so we can't keep references? That's a bit subtle, but neat.